osfoffline/utils/authentication.py

```diff
@@ -0,0 +1,189 @@
+import asyncio
+import datetime
+import furl
+import json
+import logging
+
+import aiohttp
+import bcrypt
+from PyQt5.QtWidgets import QMessageBox
+from sqlalchemy.orm.exc import NoResultFound, MultipleResultsFound
+
+from osfoffline import settings
+from osfoffline.database_manager.db import session
+from osfoffline.database_manager.models import User
+from osfoffline.database_manager.utils import save
+from osfoffline.polling_osf_manager.remote_objects import RemoteUser
+
+
+def generate_hash(password, salt=None):
+    """ Generates a password hash using `bcrypt`.
+        Number of rounds for salt generation is 12.
+
+    :return: hashed password
+    """
+    if not salt:
+        salt = bcrypt.gensalt(12)
+    return bcrypt.hashpw(
+        password.encode('utf-8'),
+        salt
+    )
+
+class AuthClient(object):
+    """Manages authorization flow """
+    def __init__(self):
+        self.failed_login = False
+
+    @asyncio.coroutine
+    def _authenticate(self, username, password):
+        """ Tries to use standard auth to authenticate and create a personal access token through APIv2
+
+            :return: personal_access_token or None
+        """
+        token_url = furl.furl(settings.API_BASE)
+        token_url.path.add('/v2/tokens/')
+        token_request_body = '{"data": {"attributes": {"name": "OSF-Offline ' + \
+            str(datetime.date.today()) + \
+            '", "scopes": "osf.full_write"}, "type": "tokens"}}'
+        headers = {'content-type': 'application/json'}
+
+        try:
+            resp = yield from aiohttp.request(method='POST', url=token_url.url, headers=headers, data=token_request_body, auth=(username, password))
+        except (aiohttp.errors.ClientTimeoutError, aiohttp.errors.ClientConnectionError, aiohttp.errors.TimeoutError):
+            # No internet connection
+            self.display_error('Unable to connect to server. Check your internet connection or try again later.')
+        except Exception as e:
+            # Invalid credentials probably, will be prompted later
+            logging.exception(e)
```

Comment:
It originally did, but when an exception was raised here that wasn't caught by the above except (manually raised), it prompted the user twice.